From a8e4e09a737cb74dbce4dda1ff6f29f4f6912dd5 Mon Sep 17 00:00:00 2001
From: zhangji <zhangji@smartisan.com>
Date: Thu, 15 Dec 2016 15:09:32 +0800
Subject: [PATCH] Fix Toast-caused crash issue

Imagine we try to add Toasts for an app, when app goes to background,
only one Toast is permitted for the logic in WindowManager, the 2nd 
Toast will cause BadTokenException, when we add the 2nd Toast again,
crash will occur.

We should not ignore this BadTokenException, call removeView when
addView is not successfully executed.

diff --git a/core/java/android/widget/Toast.java b/core/java/android/widget/Toast.java
index d807120..5addaa9b 100644
--- a/core/java/android/widget/Toast.java
+++ b/core/java/android/widget/Toast.java
@@ -496,7 +496,10 @@
                     mWM.addView(mView, mParams);
                     trySendAccessibilityEvent();
                 } catch (WindowManager.BadTokenException e) {
-                    /* ignore */
+                    // Remove mView here to cleanup from an exception
+                    // and protect future calls to handleShow
+                    e.printStackTrace();
+                    mWM.removeViewImmediate(mView);
                 }
             }
         }
